//
// See the README.md on how to compile .proto files
//

syntax = "proto3";

import "google/protobuf/any.proto";

option java_multiple_files = true;
option java_package = "io.grpc.lccsimulation";
option java_outer_classname = "LCCSimulationProto";
option objc_class_prefix = "HLW";

package lccsimulation;

enum VisObjectType
{
    // Meshes
    M_GLUCOSE = 0;
    M_ECOLI = 1;
    M_PETRI_DISH = 2;
    
    // Vis Objects
    VO_ECOLI = 3;

    // Debug Vis Objects

    // Debug Meshes
    M_SPHERE = 4;
    M_CUBE = 5;
}

//
// Main
//

service LCCSimulation 
{
    rpc Initialize (Empty) returns (InitData);
    rpc Run(Empty) returns (stream RunData);
    rpc Pause(Empty) returns (ControlSimulationResponse);
    rpc Stop(Empty) returns (ControlSimulationResponse);
}

message InitData
{
    repeated VisObjectData InitObjects = 1;
    DNAInitData DNAInit = 2;
}

message RunData
{
    SimUnitState State = 1;
    string Info = 2;
}

message ControlSimulationResponse
{
    bool Success = 1;
    string Message = 2;
}

//
// Simulation
//
message SimUnitState 
{
    float Time = 1;
    map<uint32, VisObjectData> Objects = 2;
}

message VisObjectData 
{
    uint32 ID = 1;
    VisObjectType ObjType = 2;
    Vector3 Position = 3;
    Vector3 Rotation = 4;
    Vector3 Scale = 5;
    Vector3 Color = 6;
    google.protobuf.Any VisObjectSpecificInfo = 7;
}

//
// DNA
//

message DNAAnnotation
{
    uint32 StartIndex = 1;
    uint32 EndIndex = 2;
    string Info = 3;
}

message DNAInitData 
{
    repeated uint32 DNASequence = 1;
    repeated Vector3 Points = 2;
}

//
// Common
//

message Vector3
{
    float X = 1;
    float Y = 2;
    float Z = 3;
}

message Empty {}





